File: pr-reminder-bot/listeners/actions/handle_plus_one.py
from slack_bolt import Ack
from slack_sdk import WebClient
from logging import Logger
from .block_helpers import assemble_pr_message_blocks
from database import get_pr_by_id, add_pr_to_store

def handle_plus_one(ack: Ack, body: dict, client: WebClient, logger: Logger):
    try:
        ack()

        user_id = body["user"]["id"]
        action_id = body["actions"][0]["action_id"]
        pr_id = action_id.replace("plus_one_", "")

        pr = get_pr_by_id(pr_id)

        if pr:
            # Toggle the +1 status
            if user_id in pr["reviewers"]:
                # User has already +1'd, remove the +1
                del pr["reviewers"][user_id]
                pr["reviews_received"] -= 1
            else:
                # User is +1'ing the PR, add it
                pr["reviewers"][user_id] = True
                pr["reviews_received"] += 1
            
            add_pr_to_store(pr)
            
            # Fetch the updated list of reviewers
            reviewer_names = []
            for reviewer in pr["reviewers"]:
                user_info = client.users_info(user=reviewer)
                if user_info.get("ok"):
                    real_name = user_info["user"]["real_name"]
                    reviewer_names.append(real_name)
                else:
                    logger.error(f"Failed to fetch user info for user ID {reviewer}")

            reviewers_text = ", ".join(reviewer_names) if reviewer_names else "None"
            reviews_needed = pr["reviews_needed"] - pr["reviews_received"]

            # Construct the main message text
            message_text = f"*<{pr['link']}|{pr['name']}>*\n{pr['description']}"
            if reviews_needed <= 0:
                message_text += (
                    f"\n\n:white_check_mark: Your PR was reviewed by {reviewers_text}!\n"
                    f"It’s ready to be merged.\n"
                    f"Please merge the PR and remove it from the queue, or update the request if needed."
                )
                client.chat_postMessage(
                    channel=pr['submitter_id'],
                    text= f":white_check_mark: Your PR *<{pr['link']}|{pr['name']}>* has been reviewed. Please remove it from queue or update the request.",
                    blocks= [
                        {
                            "type": "section",
                            "text": {
                                "type": "mrkdwn",
                                "text": f":white_check_mark: *Your PR* *<{pr['link']}|{pr['name']}>* *has been reviewed*.\nPlease remove it from the queue or update the request.\n<https://your-workspace.slack.com/archives/{pr['channel_id']}/p{pr['message_ts'].replace('.', '')}|View original post>"
                            }
                        }
                    ],
                    unfurl_links = False
                )
            else:
                message_text += f"\nNeeds {reviews_needed} more reviews\n\n*Reviewers:* {reviewers_text}"

            # Reconstruct the blocks with updated information
            blocks = assemble_pr_message_blocks(client, pr, user_id)

            # Update the original message with the new blocks
            client.chat_update(
                channel=pr["channel_id"],
                ts=pr["message_ts"],
                blocks=blocks,
                text=message_text
            )

    except Exception as e:
        logger.error(f"Error in handle_plus_one: {e}")

File: src/listeners/actions/HandlePlusOne.java
package prreminderbot.actions;

import com.slack.api.bolt.context.builtin.ActionContext;
import com.slack.api.methods.SlackApiException;
import com.slack.api.Slack;
import org.slf4j.Logger;
import prreminderbot.database.Database;
import prreminderbot.block_helpers.BlockHelpers;

import java.io.IOException;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;

public class HandlePlusOne {

    public static void handle(ActionRequest req, ActionContext ctx, Slack client, Logger logger) {
        try {
            ctx.ack();  // Acknowledge the action

            String userId = req.getPayload().getUser().getId();
            String actionId = req.getPayload().getActions().get(0).getActionId();
            String prId = actionId.replace("plus_one_", "");
            logger.info("User {} is attempting to +1 PR {}", userId, prId);

            Map<String, Object> pr = Database.getPrById(prId);

            if (pr != null) {
                Map<String, Boolean> reviewers = (Map<String, Boolean>) pr.get("reviewers");

                // Toggle the +1 status
                if (reviewers.containsKey(userId)) {
                    reviewers.remove(userId);
                    pr.put("reviews_received", (int) pr.get("reviews_received") - 1);
                    logger.info("User {} removed their +1 from PR {}", userId, prId);
                } else {
                    reviewers.put(userId, true);
                    pr.put("reviews_received", (int) pr.get("reviews_received") + 1);
                    logger.info("User {} added their +1 to PR {}", userId, prId);
                }

                Database.addPrToStore(pr);
                logger.info("PR {} updated in the store.", prId);

                // Fetch the updated list of reviewers
                List<String> reviewerNames = new ArrayList<>();
                for (String reviewer : reviewers.keySet()) {
                    try {
                        String realName = client.methods().usersInfo(r -> r.user(reviewer)).getUser().getRealName();
                        reviewerNames.add(realName);
                    } catch (IOException | SlackApiException e) {
                        logger.error("Failed to fetch user info for user ID {}: {}", reviewer, e.getMessage());
                    }
                }

                String reviewersText = reviewerNames.isEmpty() ? "None" : String.join(", ", reviewerNames);
                int reviewsNeeded = (int) pr.get("reviews_needed") - (int) pr.get("reviews_received");

                // Construct the main message text
                String messageText = "*" + pr.get("link") + "|" + pr.get("name") + "*\n" + pr.get("description");
                if (reviewsNeeded <= 0) {
                    messageText += (
                        "\n\n:white_check_mark: Your PR was reviewed by " + reviewersText + "!\n" +
                        "It’s ready to be merged.\n" +
                        "Please merge the PR and remove it from the queue, or update the request if needed."
                    );
                    client.methods().chatPostMessage(r -> r
                        .channel((String) pr.get("submitter_id"))
                        .text(":white_check_mark: Your PR *<" + pr.get("link") + "|" + pr.get("name") + ">* has been reviewed. Please remove it from queue or update the request.")
                        .blocks(List.of(
                            LayoutBlock.section(section -> section.text(text -> text
                                .type("mrkdwn")
                                .text(":white_check_mark: *Your PR* *<" + pr.get("link") + "|" + pr.get("name") + ">* *has been reviewed*.\n" +
                                      "Please remove it from the queue or update the request.\n" +
                                      "<https://your-workspace.slack.com/archives/" + pr.get("channel_id") +
                                      "/p" + ((String) pr.get("message_ts")).replace(".", "") + "|View original post>")
                            ))
                        ))
                        .unfurlLinks(false)
                    );
                    logger.info("PR {} has enough reviews and is ready to be merged.", prId);
                } else {
                    messageText += "\nNeeds " + reviewsNeeded + " more reviews\n\n*Reviewers:* " + reviewersText;
                }

                // Reconstruct the blocks with updated information
                List<LayoutBlock> blocks = BlockHelpers.assemblePrMessageBlocks(client, pr, userId, logger);

                // Update the original message with the new blocks
                client.methods().chatUpdate(r -> r
                    .channel((String) pr.get("channel_id"))
                    .ts((String) pr.get("message_ts"))
                    .blocks(blocks)
                    .text(messageText)
                );
                logger.info("Updated message for PR {} in channel {}.", prId, pr.get("channel_id"));

            } else {
                logger.warn("PR {} not found in the store.", prId);
            }

        } catch (IOException | SlackApiException e) {
            logger.error("Error in handle_plus_one for PR {}: {}", req.getPayload().getActions().get(0).getActionId().replace("plus_one_", ""), e.getMessage());
        }
    }
}
